src/cocaine-app/sched/__init__.py

```diff
@@ -0,0 +1,417 @@
+
+        # update job count per each type
+        self.job_count = Counter()
+
+        # update host stat
+        self.res = defaultdict(list)
+
+        errors = []
+
+        for job in active_jobs:
+            self.job_count[job.type] += 1
+            res_demand = self.convert_resource_representation(job.resources, job._involved_groups, job.type, errors)
+            self.add_to_resource_stat(res_demand, job.id)
+
+        if len(errors) != 0:
+            logger.warn("Errors during resource aggregation are {}".format(set(errors)))
+
+    def convert_resource_representation(self, resources, groups, job_type, errors=None):
+        """
+        The function takes old resources (equivalent to job.resources), "resource_limits" for each job type from config
+        and produces a new resource representation
+        :param resources:  Dict of res_type where each value is a list of node addr or tuples (node_addr, fs)
+        :param groups: List of groups ids
+        :param job_type: job_type
+        :param errors: list of strings
+        :return: a dict where keys are tuples (res_type, group) or (res_type, node_addr) or (res_type, node_addr, fs_id)
+                and values are consumption level
+        """
+        res_demand = {}
+
+        for g in groups:
+            res_demand[("Group", g)] = 100
+
+        for res_type in resources:
+            for addr in resources[res_type]:
+                if res_type != Job.RESOURCE_FS:
+                    if addr not in storage.hosts and errors:
+                        errors.append('Host with address {} is not found in storage'.format(addr))
+
+                    consumption = 100 / min(self.res_limits[job_type][res_type], 1)
+                    res_demand[(res_type, addr)] = consumption
+                else:
+                    if addr[0] not in storage.hosts and errors:
+                        errors.append('Host with address {} is not found in storage'.format(addr[0]))
+
+                    # Assume that FS is always 100% utilized
+                    res_demand[(Job.RESOURCE_FS, addr[0], addr[1])] = 100
+
+        return res_demand
+
+    def add_to_resource_stat(self, res_demand, job_id):
+        """
+        Add to saved resource representation resource consumption for the specified job. We could skip this and call
+        update_resource_stat after each iteration but this is rather expensive. So while the representation won't be
+        actual, let's keep it
+        :param res_demand: a dict where keys are tuples (res_type, node_addr\group_id, [fs_id])
+                            and values are consumption level
+        :param job_id: resource owner
+        """
+        for r in res_demand:
+            self.res[r].append((res_demand[r], job_id))
+
+    def get_resource_crossing(self, res_demand):
+        """
+        Uncovers inter-crossing of resources between running jobs and specified resources
+        :param resources: See convert_resource_representation
```

Comment:
Fixed in https://github.com/yandex/mastermind/pull/417/commits/e7a0f53468278b6dc781abea9c82cb3cb7350f4e